README.md
# tadb
## Installation
```sh
cd "~/.local/share/Steam/steamapps/common/dota 2 beta/game/dota_addons/"
git clone https://github.com/4parthy/tadb
```
## Run
Star dota 2 with in-game console enabled (launch option `-cosole`)
```sh
dota_launch_custom_game tadb touhoutd
```


changelog.txt
Update: 10 Jun @ 5:30am
Changed model files for remilia
Changed card feed bonus from mand regen to power gain for remilia.
Changed remilia's 3rd ability (buffs now apply)
Changed aya attack to use average damage
Changed flandere's 4th ability (buffs now apply)
Changed Miko's 4th ability (chance from 10% to 8.3%)
Changed Mokou's 2nd ability (keine combo increases chance by 15%)
Changed Sakuya 2nd ability blacklist (added keine)
Nerfed Seiga's 3rd ability (does half the damage now)
Changed Tenshi's attack to use average damage

Update: 13 Jun @ 9:31am
combo for sunny,luna and star
Changed Medecine's first ability (hit enemies only once per cast)
Changed Reisen ai 3rd ability (hit enemies only once per cast)
Changed Yuyuko ai 2nd ability (targeting)
Changed Utsuho(rin) ai 3rd ability (targeting)
Changed Sakuya ai 2nd, 3rd ability (targeting)
Changed Yukari ai 1st, 2nd, 4th ability (targeting)
Changed Sunny ai 2nd ability (targeting)
Added ai for Mugiyousei
Changed Keine boss (won't restore hp if already dead)
Added new model files for sakuya
Fixed saving bug in end screen
Minor changes for sanae, sunny (display skill level corectly)

Update: 15 Jun @ 9:46am
Added potrait configuration for lilly, sakuya
Minor changes to luna star sunny combo ability (mostly code cosmetics)
Changed gohei card selection (added item count check)
Nerfed Sunny damage bonus from 20/30/50 to /10/15/25 %
Nerfed Shinki (Disabled power bonus buff in endless waves)

Update: 25 Jun @ 6:50pm
Added new model files for Yukari
Added cards Suika and Yuugi
Added cards Hecatia and Junko (w/o abilities)
Added new item 1015 (w/o ability)
Changed Kaguya's 3rd ability (no more buff overwrite of other kaguya)
Added Yuugi's 3rd ability to Sakuya's 3rd ability's black list
Changed Satori's 1st ability (always re apply 2nd ability debuff)
Changed tower congfig (enabled flandre voice)
Added Suika's slow to unique slow buff list 

Update: 1 Jul @ 1:20pm 
Added new model files for Patchouli
Added abilities to Junko and Hecatia
Changed potrait configuration for many towers
Changed ai for many towers (targeting/casting)

Update: 3 Jul @ 6:41pm
Changed Junko's 3rd ability (minor fix)
Changed Sanae's 1st ability (ai targeting)
Added Junko's 1st ability buff damage
Changed ai for many towers (fixes, added junko)

Update: 3 Jul @ 4:41pm 
Added clownpiece models
Added spellicons (Keine, Luna, Star, Sunny)
Changed Hecatia planet models
Changed Hecatia abilities (finalized) 
Junko's 1st ability buff damage removes Kaguya-Boss buff.
Changed ai for Flandre, Aya, Sanae, Luna (minor adjustments)
Fixed Sanaes 4th ability combo unlock

 Update: 6 Jul @ 1:38pm 
 Changed potrait config (Added Hecatia's planets)
 Added Hecatia's 4th ability
 Changed Hecatia's abilities (damage nerf)
 Changed Luna's 1st ability (damage nerf)
 Changed ai for Flandre, Sanae, Nue, Sakuya, Yoshika (targeting)
 Changed ai for Sakuya's 2nd ability (ignore abilities with cd/mana of 1)
 Changed Lunasa's 2nd ability to unlock at 4 Stars instead 3
 
Update: 17 Mar @ 10:20am
Added Wave51
Changed players castrange from 800 to 1000 for several abilities
Removed items: 1007, 1015  
Added second level cooldown for several items
Added "shareability" to several item
Added potraits for yukari, satori and aya
Changed attack range from 800 to 100 for mugiyousei
Changed model for shanghainingyou
Changed mana regen to power buff for satori
Changed mana regen for server towers
Added Alice
Added Kokoro
Added Hina
Disabled Moniriko to be used as material in "funny" difficulty
Changed/Added difficulties
Changed Card Drawing
Changed Ayas abilities (debuff, targeting)
Changed Byakurens abilities
Changed Cirnos 2nd ability damage from 'Power * Stars 0.2' to 'Power * Stars 3' 
Changed Cirnos 3rd ability damage from '1.5 * Power' to 'Stars * Power'
Changed Cirnos 4th ability damage from '4 * Power' to 'Star * Power * 4'
Changed Cirnos 4th ability to share aoe damgae by total/4 from total/5
Changed Cirnos ExUp to apply attack time buff (6 stacks) instead of changing base attack time
Changed Cirnos ExUp to apply suwako aura to Cirnos 1s ability 
Changed Rumia ExUp to apply attack time buff (4 stacks) instead of changing base attack time
Changed Flanderes 2nd ability damage from 'Power * Illusions/2' to 'Power * 2'
Changed Flanderes 4th ability damage from 'Star * Power * 4' to 'Star * Power * 5 * 4' 
Changed Hatates 2nd ability damage from 'Power * Star / 2.5' to 'Power * Star * 2'
Changed Hecatia abilites (bugfixes, targeting)
Changed Hecatia earth/moon damge from 'Power * Star * 0.5' to 'Power * Star' [ * 2 for moon/earht type ]
Changed Hecatia triangle from 'Power * Star * 5' to Power * Star * 10'
Changed Junkos 3rd ability damage from 'Power * Star * 2' to 'Power * Star 4' (additional 20% debuff stack)
Added Junkos 4th ability (damage = Power * Star * 20)
Changed Kaguyas 1st ability damage from 'Power * Star * 0.5' 'Power * Star * 2'
Changed Kaguyas 4th ability damage from 'Power * Star / 16' to 'Power * Star'
Changed Kanakos 1st ability damage from 'Power' to from 'Power * Star'
Changed Kanakos 4th ability damage from 'Power * Star / 4' to 'Power * Star * 0.25'
Changed Kanakos 3rd/4th ability damage multiplier from 'Faith / 1000' to 'Faith / 500'
Changed Keines 1st abiility bonus damage timing
Changed Kogasa 1st ability (apply Kogasa debuff)
Changed Koishi 4th ability bonus attack damage to 'x / 0.7'
Changed Lettys 1st ability damge from 'Power' to 'Power * Star'
Changed Lunas abilities (bugifx, targeting)
Changed Lunasa abilites (combo, damage, targeting, bugfix)
Changed Lyrica abilites (combo, damage, targeting, bugfix)
Changed Marisa abilites (bugfixes, targeting)
Changed Medicine 2nd ability damage from 'Power * Star / 4' to 'Power * Star * 1.25'
Changed Meirins 1st ability added modifier
Changed Merlin abilites (combo, damage, targeting, bugfix)
Changed Minamitsu a4th ability damge from 'Power * Star * 6' to 'Power * Star * 10'

scripts/vscripts/system/damage.lua
local thtd_daiyousei_damage_bonus =
{
	[3] = 0.3,
	[4] = 0.4,
	[5] = 0.6,
}

local thtd_momiji_damage_bonus =
{
    [3] = 0.3,
    [4] = 0.4,
    [5] = 0.6,
}

local thtd_sunny_damage_bonus =
{
    [3] = 0.1,
    [4] = 0.15,
    [5] = 0.25,
}

function UnitDamageTarget(damage_table)
    local DamageTable = PassTheUnitDamageSystem(damage_table)
    damage_table = {}
	return ApplyDamage(DamageTable)
end

function PassTheUnitDamageSystem(damage_table)
    local DamageTable = clone(damage_table)   
    if DamageTable.attacker:GetUnitName() == "junko" or DamageTable.attacker:HasModifier("modifier_junko_01") then
        DamageTable.damage_type = DAMAGE_TYPE_PURE 
        if DamageTable.victim:HasModifier("modifier_junko_04_debuff") then
            DamageTable.damage = DamageTable.damage * 2           
        end
        return DamageTable
    end

    if DamageTable.damage_type == DAMAGE_TYPE_MAGICAL then
        DamageTable.damage_flags = DOTA_DAMAGE_FLAG_IGNORES_MAGIC_ARMOR
    end

    if DamageTable.attacker:FindModifierByName("modifier_item_2009_physical_penetration") ~= nil then
        if DamageTable.victim:HasModifier("modifier_item_2009_physical_penetration_effect") == false then
            DamageTable.victim:AddNewModifier(DamageTable.attacker, nil, "modifier_item_2009_physical_penetration_effect", {})
        end
    end

    if DamageTable.attacker:FindModifierByName("modifier_item_2010_magical_penetration") ~= nil then
        if DamageTable.victim:HasModifier("modifier_item_2010_magical_penetration_effect") == false then
            DamageTable.victim:AddNewModifier(DamageTable.attacker, nil, "modifier_item_2010_magical_penetration_effect", {})
        end
    end

    if DamageTable.attacker:FindModifierByName("modifier_item_2020_damage") ~= nil then       
        DamageTable.damage = DamageTable.damage + DamageTable.attacker:THTD_GetPower() * DamageTable.attacker:THTD_GetStar()
    end

    if DamageTable.victim:HasModifier("modifier_miko_01_debuff") then
        local modifier = DamageTable.victim:FindModifierByName("modifier_miko_01_debuff")
        local miko = modifier:GetCaster()        
        DamageTable.damage = DamageTable.damage + miko:THTD_GetPower()        
    end

    if DamageTable.attacker:FindModifierByName("modifier_daiyousei_03") ~= nil then
        local modifier = DamageTable.attacker:FindModifierByName("modifier_daiyousei_03")
        local daiyousei = modifier:GetCaster()
        if daiyousei ~= nil then
            if DamageTable.attacker:GetUnitName() == "cirno" then
                DamageTable.damage = DamageTable.damage * ( 1 + (thtd_daiyousei_damage_bonus[daiyousei:THTD_GetStar()] or 0) * 1.5)
            else
                DamageTable.damage = DamageTable.damage * ( 1 + (thtd_daiyousei_damage_bonus[daiyousei:THTD_GetStar()] or 0) )
            end
        end
        DamageTable.damage_type = DAMAGE_TYPE_MAGICAL 
        DamageTable.damage_flags = DOTA_DAMAGE_FLAG_IGNORES_MAGIC_ARMOR
    elseif DamageTable.attacker:FindModifierByName("modifier_momiji_02") ~= nil then
        local modifier = DamageTable.attacker:FindModifierByName("modifier_momiji_02")
        local momiji = modifier:GetCaster()
        if momiji ~= nil then
            DamageTable.damage_type = DAMAGE_TYPE_PHYSICAL 
            if DamageTable.attacker:GetUnitName() == "aya" then
                if DamageTable.victim.thtd_is_outer ~= true then
                    DamageTable.damage = DamageTable.damage * ( 1 + (thtd_momiji_damage_bonus[momiji:THTD_GetStar()] or 0) * 1.5)
                else
                    DamageTable.damage_flags = DOTA_DAMAGE_FLAG_IGNORES_PHYSICAL_ARMOR
                end
            else
                DamageTable.damage = DamageTable.damage * ( 1 + (thtd_momiji_damage_bonus[momiji:THTD_GetStar()] or 0) )
            end
        end
    elseif DamageTable.victim:HasModifier("modifier_sunny_02_debuff") then
        local modifier = DamageTable.victim:FindModifierByName("modifier_sunny_02_debuff")
        local sunny = modifier:GetCaster()
        if sunny ~= nil then
            DamageTable.damage = DamageTable.damage * ( 1 + (thtd_sunny_damage_bonus[sunny:THTD_GetStar()] or 0))
        end
        DamageTable.damage_type = DAMAGE_TYPE_PURE
    end
    return DamageTable
end

function ReturnAfterTaxDamageAfterAbility(damage_table)
    local DamageTable = PassTheUnitDamageSystem(damage_table)
    return ReturnAfterTaxDamage(DamageTable, true)
end

function ReturnAfterTaxDamage(DamageTable, Predict)
    -- 造成实际伤害后的税后伤害处理，在伤害过滤器调用，普通攻击也同样会计算
    local unit = DamageTable.attacker
    local target = DamageTable.victim    

    if unit:GetUnitName() == "yuyuko" and Predict ~= true then
        if unit.thtd_yuyuko_02_chance == nil then
            unit.thtd_yuyuko_02_chance = 5
        end
        if unit:FindAbilityByName("thtd_yuyuko_02")~=nil and RandomInt(1,100) <= unit.thtd_yuyuko_02_chance then            
            local max_damage = unit:THTD_GetPower() * unit:THTD_GetStar() * 100
            DamageTable.damage = math.min(target:GetHealth() + 100, max_damage)
            local return_damage = math.max(DamageTable.damage, 0)
            DamageTable = {}     
            return return_damage
        end
    end    

    -- 只在预估时使用，实际时在伤害过滤器中处理了
    if Predict == true and unit:GetUnitName() == "junko" or unit:HasModifier("modifier_junko_01") then
        local return_damage = math.max(DamageTable.damage, 0)
        if unit:GetUnitName() == "satori" then
            local max_damage = 3768000 * 0.5     
            return_damage = math.min(return_damage, max_damage)
        end  
        DamageTable = {}     
        return return_damage
    end

    if DamageTable.damage_type == DAMAGE_TYPE_PHYSICAL then
        if not (target.thtd_is_outer == true and unit:GetUnitName() == "aya" and unit:HasModifier("modifier_momiji_02")) then 
            local armor = target:GetPhysicalArmorValue()
            if armor > 0 then 
                if Predict ~= true then   
                    DamageTable.damage = DamageTable.damage / (1 - (0.052 * armor /(0.9 + 0.048 * armor)))  -- 7.20版本护甲计算
                end
            else
                if Predict ~= true then   
                    DamageTable.damage = DamageTable.damage / (1 - (0.052 * armor /(0.9 - 0.048 * armor)))  -- 7.20版本负护甲计算
                end
            end
        end
    end

    if target:HasModifier("modifier_bosses_hina") then
        if unit:THTD_IsTower() then            
            if DamageTable.damage > unit:THTD_GetPower()*unit:THTD_GetStar()*4 then
                DamageTable.damage = DamageTable.damage * 0.5
            end                   
        end
    end

    if target:HasModifier("modifier_bosses_kisume") then        
        DamageTable.damage = DamageTable.damage - 2000      
        if DamageTable.damage <= 0 then
            DamageTable = {}    
            return 0
        end
    end

    if unit.thtd_damage_outgoing ~= nil and unit.thtd_damage_outgoing ~= 0 then
        DamageTable.damage = DamageTable.damage * (1 + unit.thtd_damage_outgoing/100)
    end

    if target.thtd_damage_incoming ~= nil and target.thtd_damage_incoming ~= 0 then
        DamageTable.damage = DamageTable.damage * (1 + target.thtd_damage_incoming/100)
    end    

    if (target:GetHealth()/target:GetMaxHealth())>0.7 and unit:HasModifier("modifier_item_2025_damage") then
        DamageTable.damage = DamageTable.damage * 1.5
    end

    if (target:GetHealth()/target:GetMaxHealth())<0.3 and unit:HasModifier("modifier_item_2026_damage") then
        DamageTable.damage = DamageTable.damage * 1.5
    end

    if unit.equip_bonus_table ~= nil then
        if target:HasModifier("modifier_bosses_mokou")==false and Predict ~= true then
            if RandomInt(0,100) < unit.equip_bonus_table["crit_chance"] + unit.thtd_crit_chance then
                DamageTable.damage = DamageTable.damage * (1.5 + unit.equip_bonus_table["crit_damage"]/100)                       
            end
        end
        DamageTable.damage = DamageTable.damage * (1 + unit.equip_bonus_table["damage_percentage"]/100)
    end    

    if unit:HasModifier("modifier_lily_outgoing_damage") then
        DamageTable.damage = DamageTable.damage * 1.25
    end

    if DamageTable.damage_type == DAMAGE_TYPE_PHYSICAL then
        local armor = target:GetPhysicalArmorValue()       

        if unit:HasModifier("modifier_kokoro_03_buff") then
            if unit:FindModifierByName("modifier_kokoro_03_buff"):GetCaster():HasModifier("modifier_kokoro_04_buff_3") then               
                DamageTable.damage = DamageTable.damage * 1.45
            else
                DamageTable.damage = DamageTable.damage * 1.3
            end
        end

        if unit:HasModifier("modifier_utsuho_rin_buff") then
            DamageTable.damage = DamageTable.damage * 1.2
        end

        if unit:HasModifier("modifier_tenshi_03_attack_speed_buff") and Predict ~= true then
            local tenshi = unit:FindModifierByName("modifier_tenshi_03_attack_speed_buff"):GetCaster()
            if tenshi:THTD_GetStar() >= 4 and RandomInt(1,100) <= 20 then               
                DamageTable.damage = DamageTable.damage * 2
            end
        end

        if unit.thtd_physical_damage_outgoing ~= nil and unit.thtd_physical_damage_outgoing ~= 0 then
			DamageTable.damage = DamageTable.damage * (1 + unit.thtd_physical_damage_outgoing/100)
		end

        if target.thtd_physical_damage_incoming ~= nil and target.thtd_physical_damage_incoming ~= 0 then
            DamageTable.damage = DamageTable.damage * (1 + target.thtd_physical_damage_incoming/100)
        end

        if unit.equip_bonus_table ~= nil then
            DamageTable.damage = DamageTable.damage * (1 + unit.equip_bonus_table["physical_damage_percentage"]/100)
        end
        
        if armor >= 0 then 
            if unit.equip_bonus_table ~= nil then
                armor = armor * (1 - unit.equip_bonus_table["physical_penetration_percentage"]/100)
            end
            if unit.thtd_physical_penetration ~= nil and armor > 0 then
                armor = math.max(0,armor + unit.thtd_physical_penetration)
            end
            local damage_decrease = (1 - (armor * 0.04) /(1 + armor * 0.04))
            DamageTable.damage = DamageTable.damage * math.max(damage_decrease, 0.33)
        else
            local damage_decrease = (2 - 0.96^(-armor))
            DamageTable.damage = DamageTable.damage * math.min(damage_decrease, 1.66)            
        end
    end

    if DamageTable.damage_type == DAMAGE_TYPE_MAGICAL then
        local magicArmor = target:GetMagicalArmorValue() * 100

        if unit:HasModifier("modifier_patchouli_03_buff") then
            DamageTable.damage = DamageTable.damage * 1.3
        end

        if unit:HasModifier("modifier_cirno_suwako_buff") then
            DamageTable.damage = DamageTable.damage * 1.2
        end

        if unit:HasModifier("modifier_alice_shanghainingyou_crit_chance") and Predict ~= true and RandomInt(1,100) <= 20 then
            DamageTable.damage = DamageTable.damage * 2
        end
        
        if unit.thtd_magical_damage_outgoing ~= nil and unit.thtd_magical_damage_outgoing ~= 0 then
            DamageTable.damage = DamageTable.damage * (1 + unit.thtd_magical_damage_outgoing/100)
        end

        if target.thtd_magical_damage_incoming ~= nil and target.thtd_magical_damage_incoming ~= 0 then
            DamageTable.damage = DamageTable.damage * (1 + target.thtd_magical_damage_incoming/100)
        end

        if unit.equip_bonus_table ~= nil then
            DamageTable.damage = DamageTable.damage * (1 + unit.equip_bonus_table["magical_damage_percentage"]/100)
        end

        if magicArmor >= 0 then
            if unit.equip_bonus_table ~= nil then
                magicArmor = magicArmor * (1 - unit.equip_bonus_table["magical_penetration_percentage"]/100)
            end

            if unit.thtd_magical_penetration ~= nil and magicArmor > 0 then
                magicArmor = math.max(0,magicArmor + unit.thtd_magical_penetration)
            end

            local damage_decrease = (1 - (magicArmor * 0.04) /(1 + magicArmor * 0.04))           
            DamageTable.damage = DamageTable.damage * math.max(damage_decrease, 0.33)           
        else
            local damage_decrease = (2 - 0.96^(-magicArmor))
            DamageTable.damage = DamageTable.damage * math.min(damage_decrease, 1.66)           
        end
    end    

    if DamageTable.damage_type == DAMAGE_TYPE_PURE then
        if target:HasModifier("modifier_hina_01_slow_debuff") and unit:THTD_IsTower() then 
            if DamageTable.damage > unit:THTD_GetPower() * unit:THTD_GetStar() * 4 then
                DamageTable.damage = DamageTable.damage + unit:THTD_GetPower() * unit:THTD_GetStar() * 4
            elseif RandomInt(1,100) <= 40 then 
                DamageTable.damage = DamageTable.damage * 1.4
            end
        end
    end        

    local return_damage = math.max(DamageTable.damage, 0)
    if unit:GetUnitName() == "satori" then
        local max_damage = 3768000 * 0.5     
        return_damage = math.min(return_damage, max_damage)
    end
    DamageTable = {}     
    return return_damage
end

function UnitStunTarget( caster,target,stuntime)
	target:AddNewModifier(caster, nil, "modifier_stunned", {duration=stuntime})
end


local thtd_physical_penetration_buff_table = 
{
    "modifier_mystia_01_buff",
    "modifier_kaguya_03_4_buff",
}

function RefreshPhysicalPenetration(keys)
    local target = keys.target

    target.thtd_physical_penetration = 0

    for k,v in pairs(thtd_physical_penetration_buff_table) do
        local modifier = target:FindModifierByName(v)
        if modifier~=nil then
            local ability = modifier:GetAbility()
            if ability~=nil then
                target.thtd_physical_penetration = target.thtd_physical_penetration + ability:GetSpecialValueFor("penetration")
            end
        end
    end
end


local thtd_magical_penetration_buff_table = 
{
    "modifier_koakuma_02_buff",
    "modifier_byakuren_04_magic_buff",
}

function RefreshMagicalPenetration(keys)
    local target = keys.target

    target.thtd_magical_penetration = 0

    for k,v in pairs(thtd_magical_penetration_buff_table) do
        local modifier = target:FindModifierByName(v)
        if modifier~=nil then
            local ability = modifier:GetAbility()
            if ability~=nil then
                target.thtd_magical_penetration = target.thtd_magical_penetration + ability:GetSpecialValueFor("penetration")
            end
        end
    end
end


local thtd_outgoing_damage_buff_table = 
{
    "modifier_kaguya_03_3_buff",
    "modifier_merlin_01_buff",
}

function RefreshDamageOutgoingPercentage(keys)
    local target = keys.target

    target.thtd_damage_outgoing = 0

    for k,v in pairs(thtd_outgoing_damage_buff_table) do
        local modifier = target:FindModifierByName(v)
        if modifier~=nil then
            local ability = modifier:GetAbility()
            if ability~=nil then
                target.thtd_damage_outgoing = target.thtd_damage_outgoing + ability:GetSpecialValueFor("outgoing_percent")
            end
        end
    end
    
    if target:HasModifier("modifier_item_2014_damage_aura_effect") then
        target.thtd_damage_outgoing = target.thtd_damage_outgoing + 10
    end
end
